### PR TITLE
Add subscript expression support on sub-relation columns.

### PR DESCRIPTION
**Pull request by seut**
This adds support for subscript expressions on an object column of a sub-relation.
Examples: ``select a['b'] from (select a from t1)`` or
``select a['b'] from my_view`` where ``my_view`` is defined as ``select a from t1``.

### CODE REVIEW

File: common/src/main/java/io/crate/types/ObjectType.java [hunk 1]
```diff
@@ -86,6 +88,24 @@ public DataType innerType(String key) {
         return innerTypes.getOrDefault(key, UndefinedType.INSTANCE);
     }
 
+    @Nullable
+    public DataType resolveInnerType(List<String> path) {
+        if (path.size() > 0) {
```
Comment (mfussenegger) [inline common/src/main/java/io/crate/types/ObjectType.java]: Minor:

Could do an early `if (path.isEmpty() { return null; }` to reduce the indentation level of the rest of the function.

Another option would be to turn this into a non-recursive function similar to how `Maps.getByPath` is implemented. Would avoid the allocations caused by using `subList`

File: sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java [hunk 2]
```diff
@@ -619,32 +620,101 @@ Symbol resolveSubscriptSymbol(SubscriptExpression subscript, ExpressionAnalysisC
...
+            }
+            return null;
+        }
+
+        @Nullable
+        private Symbol resolveParentField(QualifiedName qualifiedName, List<String> parts) {
+            if (parts.size() == 0) {
+                return null;
+            }
+            List<String> subList = parts.subList(0, parts.size() - 1);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java]:
```suggestion
            List<String> tail = parts.subList(0, parts.size() - 1);
```

File: sql/src/main/java/io/crate/protocols/postgres/types/VarCharType.java [hunk 3]
```diff
@@ -60,8 +59,7 @@ public int writeAsText(ByteBuf buffer, @Nonnull Object value) {
 
     @Override
     protected byte[] encodeAsUTF8Text(@Nonnull Object value) {
-        assert value instanceof String : "value must be a string";
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/protocols/postgres/types/VarCharType.java]: Why can we have different types here now?

File: sql/src/main/java/io/crate/expression/scalar/SubscriptObjectFunction.java [hunk 4]
```diff
@@ -23,40 +23,51 @@
...
 public class SubscriptObjectFunction extends Scalar<Object, Map> {
 
-    public static final String NAME = "subscript_obj";
+    private static final String NAME = "subscript_obj";
     private FunctionInfo info;
 
     public static void register(ScalarFunctionModule module) {
-        module.register(NAME,
-            new BaseFunctionResolver(
-                FuncParams.builder(Param.of(ObjectType.untyped()), Param.of(StringType.INSTANCE)).build()
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/scalar/SubscriptObjectFunction.java]: Why not change the return type here to `Param.ANY` ? This would reduce the memory overhead (less entries in the function registry) and remove the need for  `allRegisteredTypes`

File: sql/src/main/java/io/crate/expression/scalar/SubscriptObjectFunction.java [hunk 5]
```diff
@@ -23,40 +23,51 @@
...
+    }
+
+    public static SubscriptObjectFunction ofReturnType(DataType returnType) {
+        return new SubscriptObjectFunction(
+            new FunctionInfo(
+                new FunctionIdent(getNameForReturnType(returnType), List.of(ObjectType.untyped(), StringType.INSTANCE)),
+                returnType));
+    }
+
+    public static String getNameForReturnType(DataType dataType) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/scalar/SubscriptObjectFunction.java]: I think this method could use some explanation. Why is this necessary?

File: sql/src/test/java/io/crate/integrationtests/SubSelectIntegrationTest.java [hunk 6]
```diff
@@ -713,4 +714,19 @@ public void testWhereSubsSelectAsClusteredByValue() {
...
+
+    public void testSubscriptOnSubSelect() {
+        execute("create table t1 (a object, c object)");
+        execute("insert into t1 (a, c) values ({ b = 1 }, { d = { e = 2 }})");
+        execute("refresh table t1");
+        execute("select a['b'], c['d']['e'] from (select * from t1) t2");
+        assertThat(printedTable(response.rows()), is("1| 2\n"));
+    }
+
+    @Ignore("Object literal inner value types are not parsed and so the fallback relying on object inner types won't work")
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/SubSelectIntegrationTest.java]: Is there a reason why they aren't parsed - if so, can we add why as a comment or so?

File: sql/src/main/java/io/crate/protocols/postgres/types/VarCharType.java (hunk 3 above)
Comment (seut) [reply sql/src/main/java/io/crate/protocols/postgres/types/VarCharType.java]: Because the UndefinedType is mapped to a PG VarCharType and thus not only strings will end up here.

File: sql/src/main/java/io/crate/expression/scalar/SubscriptObjectFunction.java (hunk 4 above)
Comment (seut) [reply sql/src/main/java/io/crate/expression/scalar/SubscriptObjectFunction.java]: Hm don't get your point. This is all about the returnType of the function and not the arguments.
This ensures that the correct data type is used for the according java type which is used on streaming, encoding etc.

File: sql/src/test/java/io/crate/integrationtests/SubSelectIntegrationTest.java (hunk 6 above)
Comment (seut) [reply sql/src/test/java/io/crate/integrationtests/SubSelectIntegrationTest.java]: We discussed and agreed that we don't want to implement parsing of user-given object literals currently. If you changed your mind, I'm open for re-discussion.

File: sql/src/main/java/io/crate/expression/scalar/SubscriptObjectFunction.java (hunk 5 above)
Comment (seut) [reply sql/src/main/java/io/crate/expression/scalar/SubscriptObjectFunction.java]: See my previous comment. Our function registration does not take the function return type into account. I solved this by registering the same function with a different return type under different names. Any other suggestion is appreciated.

File: common/src/main/java/io/crate/types/ObjectType.java (hunk 1 above)
Comment (seut) [reply common/src/main/java/io/crate/types/ObjectType.java]: +1 for an non-recursive function, will try to eliminate that.

File: sql/src/main/java/io/crate/expression/scalar/SubscriptObjectFunction.java (hunk 5 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/expression/scalar/SubscriptObjectFunction.java]: Sorry I think I wasn't clear - I think this aspect of using the type name within the function should be documented (explaining why it is necessary) in the code somewhere. Either on the class level or here on this method.

File: sql/src/main/java/io/crate/protocols/postgres/types/VarCharType.java (hunk 3 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/protocols/postgres/types/VarCharType.java]: Is `toString` safe for all types ? E.g. what if it is an array or an object itself?

File: sql/src/test/java/io/crate/integrationtests/SubSelectIntegrationTest.java (hunk 6 above)
Comment (mfussenegger) [reply sql/src/test/java/io/crate/integrationtests/SubSelectIntegrationTest.java]: Ah right. Maybe clarify this with something like 

```suggestion
    @Ignore("We haven't added inner types for object literals, so this test is expected to fail")
```

File: sql/src/main/java/io/crate/protocols/postgres/types/VarCharType.java (hunk 3 above)
Comment (seut) [reply sql/src/main/java/io/crate/protocols/postgres/types/VarCharType.java]: Yes right, I was also thinking that this should be fixed properly maybe in a dedicated commit/PR as it is not really related to this one.

File: sql/src/main/java/io/crate/expression/scalar/SubscriptObjectFunction.java (hunk 5 above)
Comment (seut) [reply sql/src/main/java/io/crate/expression/scalar/SubscriptObjectFunction.java]: Ah ok, right will do.

Comment (seut) [reply sql/src/main/java/io/crate/expression/scalar/SubscriptObjectFunction.java]: I've added class level documentation.

File: sql/src/main/java/io/crate/protocols/postgres/types/VarCharType.java (hunk 3 above)
Comment (seut) [reply sql/src/main/java/io/crate/protocols/postgres/types/VarCharType.java]: Calling `toString` is safe on any instance but it won't result in useful content. As this is more a fallback than the usual case, I'm not sure how far we wanna go here with arrays/list/maps/objects. What do you think?

Comment (seut) [reply sql/src/main/java/io/crate/protocols/postgres/types/VarCharType.java]: I've removed this change as it is not related to this PR.

**Comment by seut** [comment]
@mfussenegger I've pushed fixup commits which addressed all comments but accidentally squashed them by master rebase :-( sorry

File: sql/src/main/java/io/crate/expression/scalar/SubscriptObjectFunction.java [hunk 7]
```diff
@@ -31,32 +31,75 @@
...
+            } else {
+                returnTypeName = dataType.getName();
             }
-        );
+            name += "_" + returnTypeName;
+        }
+        return name;
+    }
+
+    public static <T> List<T> buildListOfRootAndPath(T rootElement, List<String> path, Function<String, T> convert) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/scalar/SubscriptObjectFunction.java]: Maybe call this `mapTail(O head, List<I> tail, Function<I, O> mapper)`  and put it into `Lists2` ?

Currently it is somehow generic and yet at the same time very specific. (`rootElement`, `path` imply a concrete use-case). I also wasn't able to deduct from the method name alone what exactly it does.

File: sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java [hunk 8]
```diff
@@ -619,30 +622,97 @@ Symbol resolveSubscriptSymbol(SubscriptExpression subscript, ExpressionAnalysisC
...
 
-        private Symbol createSubscript(Symbol name, List<String> parts, ExpressionAnalysisContext context) {
-            Symbol subscript = name;
-            for (int i = 0; i < parts.size(); i++) {
-                subscript = allocateFunction(
-                    SubscriptObjectFunction.NAME, ImmutableList.of(subscript, Literal.of(parts.get(i))), context);
+        private Symbol createSubscript(Symbol symbol, List<String> parts, ExpressionAnalysisContext context) {
+            DataType symbolType = symbol.valueType();
+            List<Symbol> arguments = buildListOfRootAndPath(symbol, parts, Literal::of);
+            List<DataType> argumentTypes = buildListOfRootAndPath(symbolType, parts, s -> DataTypes.STRING);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java]:
```suggestion
            List<DataType> argumentTypes = buildListOfRootAndPath(symbolType, parts, ignored -> DataTypes.STRING);
```

or 

```suggestion
            List<DataType> argumentTypes = Symbols.typeView(arguments);
```
